### PR TITLE
[gen4 planner] Better handling of projections in operators

### PR DESCRIPTION
**Pull request by systay**
## Description
This pull request introduces significant changes to the operator planning model, enhancing its structure and functionality. The updated model consists of the following steps:

1. Construct a logical operator tree from the AST: This tree remains unoptimized, with all components of SELECT expressions, ORDER BY, GROUP BY, and LIMIT packed into a Horizon operator.
2. Expand the leaves of the operator tree: Determine the join order of the routes.
3. Attempt to push down Horizon operations beneath the routes: If successful, the AST fields from the input query can be directly copied to the query sent to the tablet.
4. Expand the horizons into smaller components (Project, Aggregate, Sort, Limit, etc.): Try to push these components down beneath the routes.

The main advantage of this new model is that after each Horizon push or expansion, we have a functional query plan. This allows us to view intermediate stages as complete plans, which can be beneficial for debugging. For instance, we can execute every intermediate step to identify which optimization introduces an undesirable result.

## Related Issue(s)
Tracking issue: https://github.com/vitessio/vitess/issues/11626

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/derived.go [hunk 1]
```diff
@@ -218,3 +220,12 @@ func addToIntSlice(columnOffset []int, valToAdd int) ([]int, int) {
 	columnOffset = append(columnOffset, valToAdd)
 	return columnOffset, len(columnOffset) - 1
 }
+
+// TODO: REMOVE
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/derived.go:224]: this should be cleaned up before we merge this PR

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 2]
```diff
@@ -59,17 +61,70 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
...
+	})
+
+	return &projection{
+		source:      src,
+		columnNames: op.ColumnNames,
+		columns:     expressions,
+	}, nil
+}
+
+func simplifyPlan(op *operators.Projection, cols []int, src logicalPlan) (logicalPlan, error) {
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operator_transformers.go:98]: nit: `simplifyPlan` is a very generic name. Could we call it something more explanatory like `convertToSimpleProjection`, etc.

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 3]
```diff
@@ -159,57 +182,113 @@ func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.
...
 		}
-		for i, lhsExpr := range lhsExprs {
+	default:
+		return JoinColumn{}, vterrors.VT13002(sqlparser.String(e))
+	}
+
+	return
+}
+
+func (a *ApplyJoin) AddColumn(ctx *plancontext.PlanningContext, expr *sqlparser.AliasedExpr) (ops.Operator, int, error) {
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:215]: nit: Some of the exported functions don't have comments.

File: go/vt/vtgate/planbuilder/operators/expressions.go [hunk 4]
```diff
@@ -57,8 +57,24 @@ func BreakExpressionInLHSandRHS(
...
+		return JoinColumn{}, err
 	}
 	ctx.JoinPredicates[expr] = append(ctx.JoinPredicates[expr], rewrittenExpr)
+	col.RHSExpr = rewrittenExpr
 	return
 }
+
+// BreakExpressionInLHSandRHSOld takes an expression and
+// extracts the parts that are coming from one of the sides into `ColName`s that are needed
+func BreakExpressionInLHSandRHSOld(
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/expressions.go:69]: Why is this required? Couldn't we change the places using this to expect a JoinColumn instead?

File: go/vt/vtgate/planbuilder/operators/horizon.go [hunk 5]
```diff
@@ -60,3 +74,17 @@ func (h *Horizon) Clone(inputs []ops.Operator) ops.Operator {
 func (h *Horizon) Inputs() []ops.Operator {
 	return []ops.Operator{h.Source}
 }
+
+// SetInputs implements the Operator interface
+func (h *Horizon) SetInputs(ops []ops.Operator) {
+	h.Source = ops[0]
+}
+
+// TODO: REMOVE
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/horizon.go:83]: cleanup

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 2 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operator_transformers.go:98]: I agree with Manan's statement above, `simplifyPlan` is somewhat generic.

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 6]
```diff
@@ -59,17 +61,70 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
...
+	// all columns are just passing through something from below
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operator_transformers.go:99]: I might misunderstand this comment, but I think it is misplaced? The code below simply return an error.

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 7]
```diff
@@ -59,17 +61,70 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
...
+// their corresponding index values. It returns true if all elements match
+// their indices, and false otherwise.
+func elementsMatchIndices(in []int) bool {
+	for idx, val := range in {
+		if val != idx {
+			return false
+		}
+	}
+	return true
+}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operator_transformers.go:126]: I think it would be nice to have some sort of `utils` package where we can put utility function like this one. Similar to the `utils` end-to-end package.

Comment (frouioui) [reply go/vt/vtgate/planbuilder/operator_transformers.go:126]: This would force us to reuse code we already have and, not saying this is duplicated, but it would avoid duplication in the future and ease maintenance / testability

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 2 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operator_transformers.go:98]: we previously used `transformSimpleProjection` this is also a good name

File: go/vt/vtgate/planbuilder/operators/SQL_builder.go [hunk 8]
```diff
@@ -312,6 +317,20 @@ func buildQuery(op ops.Operator, qb *queryBuilder) error {
...
+		}
+
+		qb.clearProjections()
+		for i, column := range op.Columns {
+			ae := &sqlparser.AliasedExpr{Expr: column.GetExpr()}
+			if op.ColumnNames[i] != "" {
+				ae.As = sqlparser.NewIdentifierCI(op.ColumnNames[i])
+			}
+			qb.addProjection(ae)
+		}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/SQL_builder.go:333]: nit: we could also move this to its own little function - that would take us one step closer to having all query serving `switch` easily readable :)

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 9]
```diff
@@ -32,26 +32,50 @@ import (
...
-
 	// LeftJoin will be true in the case of an outer join
 	LeftJoin bool
 
 	// JoinCols are the columns from the LHS used for the join.
 	// These are the same columns pushed on the LHS that are now used in the Vars field
 	LHSColumns []*sqlparser.ColName
 
+	// Before offset planning
+
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:43]: these comments and the organisation are nice!

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 10]
```diff
@@ -32,26 +32,50 @@ import (
...
+//   - Pure right
+//     We are projecting a column that comes from the right. The LHSExprs will be empty for these
+//   - Mix of data from left and right
+//     Here we need to transmit columns from the LHS to the RHS,
+//     so they can be used for the result of this expression that is using data from both sides.
+//     All fields will be used for these
+type JoinColumn struct {
+	Original sqlparser.Expr // this is the original expression being passed through
+	BvNames  []string       // the BvNames and LHSCols line up
+	LHSExprs []sqlparser.Expr
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:75]: nit, I think you meant:

```suggestion
	BvNames  []string       // the BvNames and LHSExprs line up
	LHSExprs []sqlparser.Expr
```

File: go/vt/vtgate/planbuilder/operators/apply_join.go (hunk 3 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/apply_join.go:215]: I think it's _okay_ for interface implementation as long as the interface definition contains a comment for this method, which it does here

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 11]
```diff
@@ -159,57 +182,113 @@ func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.
...
+		// Read the type description for JoinColumn to understand the following code
+		for i, lhsExpr := range col.LHSExprs {
 			offset, err := a.pushColLeft(ctx, aeWrap(lhsExpr))
 			if err != nil {
-				return nil, 0, err
+				return err
+			}
+			if col.RHSExpr == nil {
+				// if we don't have a RHS expr, it means that this is a pure LHS expression
+				a.addOffset(-offset - 1)
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:237]: it's not exactly clear why we are doing `-offset` here. I am guessing it's because offsets for `LHS` are going to be negative, and we want to turn it into a positive here, but I am not really sure.

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 12]
```diff
@@ -159,57 +182,113 @@ func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.
...
+				return err
 			}
-			a.Vars[bvNames[i]] = offset
+			a.addOffset(offset + 1)
 		}
-		offset, err := a.pushColRight(ctx, aeWrap(rhsExpr))
+	}
+
+	for _, col := range a.JoinPredicates {
+		var err error
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:252]: nit: we could define this outside of the loop to avoid or even as a named return param

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 13]
```diff
@@ -159,57 +182,113 @@ func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.
...
+			offset, err := a.pushColLeft(ctx, aeWrap(lhsExpr))
+			if err != nil {
+				return err
+			}
+			a.Vars[col.BvNames[i]] = offset
+		}
+		lhsColumns := slices2.Map(col.LHSExprs, func(from sqlparser.Expr) *sqlparser.ColName {
+			col, ok := from.(*sqlparser.ColName)
+			if !ok {
+				// todo: there is no good reason to keep this limitation around
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:263]: What would it take to support more than `ColName`? My intuition is that we would first need to check if the expression is something the EvalEngine can handle

File: go/vt/vtgate/planbuilder/operators/derived.go (hunk 1 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/derived.go:224]: Let's clean it up :p

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 14]
```diff
@@ -19,76 +19,179 @@ package operators
...
+	ctx *plancontext.PlanningContext,
+	p *Projection,
+) (ops.Operator, rewrite.ApplyResult, error) {
+	switch src := p.Source.(type) {
+	case *Route:
+		op, err := rewrite.Swap(p, src)
+		if err != nil {
+			return nil, false, err
+		}
+		return op, rewrite.NewTree, nil
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:79]: let's extract this to `pushDownProjectionInRoute` 😄

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 15]
```diff
@@ -19,76 +19,179 @@ package operators
...
+		src.LHS = lhsProj
+	}
+	if len(rhsColumns) > 0 {
+		rhsProj, err := createProjection(src.RHS)
+		if err != nil {
+			return nil, false, err
+		}
+		rhsProj.ColumnNames = rhsNames
+		rhsProj.Columns = rhsColumns
+		src.RHS = rhsProj
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:176]: It would be great to extract this into a single method that can be reused for both `lhsCols` and `rhsColumns`, i think it would be easier to maintain :) no strong opinion tho

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 16]
```diff
@@ -19,76 +19,179 @@ package operators
...
+	var (
+		// idx is the current column index in the Projection's columns.
+		// in is the current column's ProjExpr.
+		idx int
+		in  ProjExpr
 
-	sel, isSel := in.Select.(*sqlparser.Select)
+		// lhsCols and rhsColumns store the columns to be added to the left and right child projections.
+		// lhsNames and rhsNames store the corresponding column names.
+		lhsCols, rhsColumns []ProjExpr
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:117]: nit: I think it would be `lhsCols` and `rhsCols` for both sides? unless there is some naming convention that i am missing for ApplyJoins

```suggestion
		// lhsCols and rhsCols store the columns to be added to the left and right child projections.
		// lhsNames and rhsNames store the corresponding column names.
		lhsCols, rhsCols []ProjExpr
```

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 17]
```diff
@@ -97,118 +200,60 @@ func planHorizon(ctx *plancontext.PlanningContext, in *Horizon, isRoot bool) (op
...
+
+	return expandHorizon(qp, in)
+}
+
+// horizonLike should be removed. we should use Horizon for both these cases
+type horizonLike interface {
+	ops.Operator
+	selectStatement() sqlparser.SelectStatement
+	src() ops.Operator
 }
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:217]: `horizonLike` should be removed from the codebase before we merge this or from the operator tree?

File: go/vt/vtgate/planbuilder/operators/operator.go [hunk 18]
```diff
@@ -51,44 +51,111 @@ type (
...
+	// next we use the original Horizon to make sure that the output columns line up with what the user asked for
+	// in the future, we'll tidy up the results. for now, we are just failing these queries and going back to the
+	// old horizon planning instead
+	horizon := op.(*Horizon)
+
+	sel := sqlparser.GetFirstSelect(horizon.Select)
+	cols, err := output.GetColumns()
+	if err != nil {
+		return err
+	}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/operator.go:132]: nit picking here: fail early

```suggestion
	cols, err := output.GetColumns()
	if err != nil {
		return err
	}
	
	horizon := op.(*Horizon)
	sel := sqlparser.GetFirstSelect(horizon.Select)
```

File: go/vt/vtgate/planbuilder/operators/projection.go [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+type (
+	// Projection is used when we need to evaluate expressions on the vtgate
+	// It uses the evalengine to accomplish its goal
+	Projection struct {
+		Source      ops.Operator
+		ColumnNames []string
+		Columns     []ProjExpr
+	}
+	ProjExpr interface {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/projection.go:36]: nit: new line

**Review by frouioui — APPROVED** [review]
Left several comments which are mostly nit picks / questions; Nothing urgent to fix to report or anything that seemed broken to me. Really cool PR, excited to land this!

File: go/vt/vtgate/planbuilder/operators/apply_join.go (hunk 13 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/apply_join.go:263]: the comparison is usually done by sending down the value from the LHS to the RHS. We don't actually evaluate these expressions in the vtgate, so evalengine does not come into play. I don't know of any good reason to limit like this.

File: go/vt/vtgate/planbuilder/operators/derived.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/derived.go:224]: we are fixing this here: https://github.com/vitessio/vitess/pull/12887

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go (hunk 17 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go]: Seems like https://github.com/vitessio/vitess/pull/12887 is closed and will be ported over to this Pull Request.

File: go/vt/vtgate/planbuilder/operators/apply_join.go (hunk 11 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/apply_join.go:237]: Makes sense. Now I remember :)

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go (hunk 17 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go:217]: yeah, and now we regretted this decision. we still want to remove derived, but not now. we'll do it after we are done with porting over the horizon planning to operators
